sympy/core/basic.py

```diff
@@ -503,11 +503,11 @@ def atoms(self, *types):
         if types:
             types = tuple(
                 [t if isinstance(t, type) else type(t) for t in types])
-        else:
-            types = (Atom,)
         result = set()
         for expr in preorder_traversal(self):
-            if isinstance(expr, types):
+            if not (types or expr.args):
+                result.add(expr)
+            elif isinstance(expr, types):
```

Comment:
I think it would be clearer to write:
```python
nodes = preorder_traversal(self)
if types:
    result = {node for node in nodes if isinstance(node, types)}
else:
    result = {node for node in nodes if not node.args}
```

Reply:
@oscarbenjamin sure thing.